File: client/node/mcp-client.js
const axios = require("axios");
const fs = require("fs");

class MCPClient {
  constructor(options = {}) {
    this.url = options.url || process.env.CENTRAL_MCP_SERVER_URL || null;
    this.token = options.token || process.env.CENTRAL_MCP_CLIENT_TOKEN || null;
    this.cacheTtl = options.cacheTtl || 30000; // ms
    this._cache = { data: null, ts: 0 };
  }

  _now() {
    return Date.now();
  }

  async _fetchRemote() {
    if (!this.url) throw new Error("MCP URL not configured");
    const headers = {};
    if (this.token) headers["Authorization"] = `Bearer ${this.token}`;
    const resp = await axios.get(`${this.url.replace(/\/$/, "")}/context`, {
      headers,
      timeout: 5000,
    });
    return resp.data;
  }

  _readLocal() {
    try {
      const content = fs.readFileSync("C:/central-mcp-config.json", "utf8");
      return JSON.parse(content);
    } catch {
      return null;
    }
  }

  async getAll() {
    // use cache
    if (this._cache.data && this._now() - this._cache.ts < this.cacheTtl)
      return this._cache.data;

    // 1) env vars -> build object of any CENTRAL_* envs
    const envObj = {};
    Object.keys(process.env).forEach((k) => {
      if (k.startsWith("CENTRAL_")) envObj[k.toLowerCase()] = process.env[k];
    });
    if (Object.keys(envObj).length) {
      this._cache = { data: envObj, ts: this._now() };
      return envObj;
    }

    // 2) local file
    const local = this._readLocal();
    if (local) {
      this._cache = { data: local, ts: this._now() };
      return local;
    }

    // 3) remote
    try {
      const remote = await this._fetchRemote();
      this._cache = { data: remote, ts: this._now() };
      return remote;
    } catch (e) {
      throw new Error(
        "Failed to load config from env/local/remote: " + e.message
      );
    }
  }

  async get(key) {
    const all = await this.getAll();
    return all ? all[key] : undefined;
  }

  async getSecret(name) {
    // 1) direct env checks
    if (process.env[name]) return process.env[name];
    const up = name.toUpperCase();
    if (process.env[up]) return process.env[up];

    // 2) local file
    try {
      const content = fs.readFileSync("C:/central-mcp-config.json", "utf8");
      const cfg = JSON.parse(content);
      if (
        cfg.secrets &&
        Object.prototype.hasOwnProperty.call(cfg.secrets, name)
      )
        return cfg.secrets[name];
      if (Object.prototype.hasOwnProperty.call(cfg, name)) return cfg[name];
    } catch {
      // try workspace file
      try {
        const content = fs.readFileSync(
          require("path").join(
            __dirname,
            "..",
            "..",
            "central-mcp-config.json"
          ),
          "utf8"
        );
        const cfg = JSON.parse(content);
        if (
          cfg.secrets &&
          Object.prototype.hasOwnProperty.call(cfg.secrets, name)
        )
          return cfg.secrets[name];
        if (Object.prototype.hasOwnProperty.call(cfg, name)) return cfg[name];
      } catch {
        // ignore
      }
    }

    // 3) remote via /secrets (try direct token first)
    if (!this.url) throw new Error("MCP URL not configured");
    const base = this.url.replace(/\/$/, "");
    const headers = {};
    if (this.token) headers["Authorization"] = `Bearer ${this.token}`;
    try {
      if (this.token) {
        const resp = await axios.get(
          `${base}/secrets/${encodeURIComponent(name)}`,
          { headers, timeout: 5000 }
        );
        if (resp && resp.data && resp.data.value) return resp.data.value;
      }
    } catch {
      // fallthrough to token exchange
    }

    // obtain short-lived JWT
    try {
      if (this.token) {
        const t = await axios.post(`${base}/token`, null, {
          headers,
          timeout: 5000,
        });
        const access = t.data && t.data.access_token;
        if (access) {
          const resp2 = await axios.get(
            `${base}/secrets/${encodeURIComponent(name)}`,
            { headers: { Authorization: `Bearer ${access}` }, timeout: 5000 }
          );
          if (resp2 && resp2.data && resp2.data.value) return resp2.data.value;
        }
      }
    } catch {
      // ignore
    }

    return undefined;
  }
}

module.exports = MCPClient;

File: client/python/mcp_client.py
import os
import time
import json
import requests
from typing import Optional


class MCPClient:
    def __init__(
        self,
        url: Optional[str] = None,
        token: Optional[str] = None,
        cache_ttl: int = 30,
    ):
        self.url = url or os.environ.get("CENTRAL_MCP_SERVER_URL")
        self.token = token or os.environ.get("CENTRAL_MCP_CLIENT_TOKEN")
        self.cache_ttl = cache_ttl
        self._cache = {"data": None, "ts": 0}

    def _now(self):
        return int(time.time())

    def _read_local(self):
        # prefer absolute C: path, then workspace file
        candidates = [
            r"C:\central-mcp-config.json",
            os.path.join(
                os.path.dirname(__file__), "..", "..", "central-mcp-config.json"
            ),
        ]
        for p in candidates:
            try:
                if os.path.exists(p):
                    with open(p, "r", encoding="utf-8") as f:
                        return json.load(f)
            except Exception:
                continue
        return None

    def _fetch_remote(self):
        if not self.url:
            raise RuntimeError("MCP URL not configured")
        headers = {}
        if self.token:
            headers["Authorization"] = f"Bearer {self.token}"
        resp = requests.get(
            self.url.rstrip("/") + "/context", headers=headers, timeout=5
        )
        resp.raise_for_status()
        return resp.json()

    def get_all(self):
        if self._cache["data"] and (self._now() - self._cache["ts"]) < self.cache_ttl:
            return self._cache["data"]
        # 1) env vars starting with CENTRAL_
        env_obj = {k: v for k, v in os.environ.items() if k.startswith("CENTRAL_")}
        if env_obj:
            self._cache = {"data": env_obj, "ts": self._now()}
            return env_obj
        # 2) local file
        local = self._read_local()
        if local:
            self._cache = {"data": local, "ts": self._now()}
            return local
        # 3) remote
        remote = self._fetch_remote()
        self._cache = {"data": remote, "ts": self._now()}
        return remote

    def get(self, key: str):
        allcfg = self.get_all()
        return allcfg.get(key) if allcfg else None

    def get_secret(self, name: str):
        # 1) check environment directly
        if name in os.environ:
            return os.environ[name]
        env_key = name.upper()
        if env_key in os.environ:
            return os.environ[env_key]

        # 2) check local config file
        local = self._read_local()
        if local:
            if isinstance(local, dict):
                if "secrets" in local and name in local["secrets"]:
                    return local["secrets"][name]
                if name in local:
                    return local[name]

        # 3) try remote: first try direct token (server token) then try obtaining short-lived JWT
        if not self.url:
            return None
        headers = {}
        # try direct token (may be server token)
        if self.token:
            headers["Authorization"] = f"Bearer {self.token}"
            try:
                r = requests.get(
                    self.url.rstrip("/") + f"/secrets/{name}",
                    headers=headers,
                    timeout=5,
                )
                if r.status_code == 200:
                    return r.json().get("value")
            except Exception:
                pass

        # try obtain short-lived JWT
        try:
            if self.token:
                hdr = {"Authorization": f"Bearer {self.token}"}
                t = requests.post(
                    self.url.rstrip("/") + "/token", headers=hdr, timeout=5
                )
                if t.status_code == 200:
                    token = t.json().get("access_token")
                    if token:
                        hdr2 = {"Authorization": f"Bearer {token}"}
                        r2 = requests.get(
                            self.url.rstrip("/") + f"/secrets/{name}",
                            headers=hdr2,
                            timeout=5,
                        )
                        if r2.status_code == 200:
                            return r2.json().get("value")
        except Exception:
            pass

        return None


if __name__ == "__main__":
    c = MCPClient()
    print("MCP URL from client:", c.url)
    print("Client token exists:", bool(c.token))
    print("All config:", c.get_all())
    print("Secret mySecretKey via env/local/remote:", c.get("mySecretKey"))
